arch/riscv/core/smp.c

```diff
@@ -28,6 +28,10 @@ extern void __start(void);
 void soc_interrupt_init(void);
 #endif
 
+#ifdef CONFIG_RISCV_SOC_HAS_CUSTOM_PER_CORE_INIT
+void z_soc_per_core_init(void);
```

Comment:
lets not add more `z_soc...` code into the tree, just drop the `z_ ` from the name, this can be confusing as  it describes something internal to zephyr that should not be used outside of zephyr, what you are doing here is defining a new "interface", so it is an API somehow.